projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,343 @@
+---
+title: Health ping data analysis (Nightly)
+authors:
+- kateu
+tags:
+- health ping
+created_at: 2017-08-10 00:00:00
+updated_at: 2017-08-10 17:03:51.444201
+tldr: Health ping validation, clients per day, pings per day, clients per ping per
+  day, health ping day clients vs DAU.
+thumbnail: images/output_8_0.png
+---
+```python
+import ujson as json
+import matplotlib.pyplot as plt
+import pandas as pd
+import numpy as np
+import plotly.plotly as py
+
+from plotly.graph_objs import *
+from moztelemetry import get_pings_properties, get_one_ping_per_client
+from moztelemetry.dataset import Dataset
+get_ipython().magic(u'matplotlib inline')
+```
+
+```python
+pings = Dataset.from_source("telemetry").where(docType='health').where(appUpdateChannel="nightly").records(sc, sample=1)
+cachedData = get_pings_properties(pings, ["payload/pingDiscardedForSize", "payload/sendFailure", 
+                                             "clientId", "meta/submissionDate", "payload/os", "payload/reason"]).cache()
+```
+    fetching 126.43106MB in 34883 files...
+
+
+# Compute failures stats for each failure: sendFailre, discardedForSize. 
+* for sendFailure stats include health ping count per failure type
+* for discardedForSize stats include health ping count per ping type
+
+
+```python
+def aggregateFailures(first, second):
+    if first is None:
+        return second
+    if second is None:
+        return first
+    
+    res = first
+    for k, v in second.items():
+        if isinstance(v, int):
+                if k in res:
+                    res[k] += v
+                else: 
+                    res[k] = v;
+    return res
+```
+
+```python
+# return array of pairs [(failureName, {failureStatistic: count, ....}), ...]
+# e.g. [(discardedForSize, {"main": 3, "crash": 5}), (sendFailure, {"timeout" : 34})]
+def getFailuresStatPerFailureName(pings, failureNames):
+    def reduceFailure(failureName):
+        return pings.map(lambda p: p[failureName]).reduce(aggregateFailures)
+    
+    healthPingFailuresStat = map(reduceFailure, failuresNames)
+    return zip(failuresNames, healthPingFailuresStat)
+    
+```
+
+```python
+failuresNames = ["payload/pingDiscardedForSize", "payload/sendFailure"]
+```
+
+```python
+failuresStat = getFailuresStatPerFailureName(cachedData, failuresNames)
+```
+
+```python
+for fs in failuresStat:
+    plt.title(fs[0])
+    plt.bar(range(len(fs[1])), fs[1].values(), align='center')
+    plt.xticks(range(len(fs[1])), fs[1].keys())
+    plt.show()
+    
+```
+
+
+![png](images/output_8_0.png)
```

Comment:
Right, that would add some data if we need it.
This is not anything we need to do right now anyway, but it will be a good future improvement for the health ping.